CORDA-3490 Start node without running all checkpointed flows

williamvigorr3 opened the pull request:
Add Statemachine support API to support pausing and resuming flow.

As a part of this PR the state machine supports pausing a Hospitalised Flow. Alternatively a flow can be paused by shutting down the state machine and individually selecting flows which should be paused when the state machine restarts.

We also need to decide how a user will interact with the Statemachine to pause and resume flows (currently I have added an internal RPC Opp and the StateMachineManager.StartMode to the node configuration options). This has let me write some tests.

williamvigorr3 committed:
Refactored function for clarity

williamvigorr3 committed:
Modified statemachine to store checkpoints without deserializing flow state

This should be memory efficient when pausing flows.

williamvigorr3 committed:
Finished implementing Map of paused flows

This needs testing

williamvigorr3 committed:
Added unpause node to statemachine API.

Also lots of minor fixes. This is all untested so far.

williamvigorr3 committed:
Minor rename

williamvigorr3 committed:
Support pausing hospitalised flows.

williamvigorr3 committed:
Check in database for paused flow.

williamvigorr3 committed:
Implemented statemachine startup mode.

williamvigorr3 committed:
Added Integration test of Safe start mode

This is just a skeleton for now. Eventually we want to check
that we can send messages between nodes when a flow is paused.
These should get added to a queue and processed if the flow gets
resumed.

williamvigorr3 committed:
Added queue of unprocessed events for paused flows

Added Integration test to check this works.

williamvigorr3 committed:
Make Detekt pass again :)

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,95 @@ class SingleThreadedStateMachineManager(
 
     private class Flow(val fiber: FlowStateMachineImpl<*>, val resultFuture: OpenFuture<Any?>)
 
+    private inner class NonResidentFlow(val id: StateMachineRunId,
```
If possible i think we should try and extract some code out of the single threaded manager but that can come after we get everything working.

lankydan replied:
I assume it is an inner class due to referencing things like the `actionExecutor` and `transitionExecutor`?

williamvigorr3 replied:
Yes that is correct. I agree we should try and pull this out into its own file. It will be easier to refactor this now (while how it works is fresh in my head).

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -144,6 +235,12 @@ class SingleThreadedStateMachineManager(
         fiberDeserializationChecker?.start(checkpointSerializationContext)
+        when (startMode) {
+            StateMachineManager.StartMode.ResumeAll -> {}
+            StateMachineManager.StartMode.Safe -> markAllFlowsAsPaused()
```
`Safe` and `PauseHospitalised` should be the same. From a user's perspective there should be no `RUNNABLE` flows when a node restarts. Obviously we know that is not true, I don't think we need to make it more confusing by distinguishing between them. Therefore I think there is either start all or start none.

williamvigorr3 replied:
I will remove the `PauseHospitalised` mode (which just pauses hospitalised flows at the moment). This doesn't actually work as we are never setting the status to hospitalized.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -338,17 +501,23 @@ class SingleThreadedStateMachineManager(
+                // If a flow is added before start() then don't attempt to restore it
+                val checkpoint = tryDeserializeCheckpoint(serializedCheckpoint, id) ?: return@mapNotNull null
+                id to NonResidentFlow(id, checkpoint, true, false, null)
+            }.toList().map {pair -> pair.first to pair.second}.toMap()
```
I think you can just do `toMap` and remove the `map`, or something similar to that.

williamvigorr3 replied:
This should be fixed now.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -373,14 +542,12 @@ class SingleThreadedStateMachineManager(
                 }
+
+                val checkpoint = tryDeserializeCheckpoint(serializedCheckpoint, flowId) ?: return
+                val nonResidentFlow = NonResidentFlow(flowId, checkpoint, false, false, null)
```
I find the mentioning of `NonResidentFlow` confusing here. I understand that we are calling `NonResidentFlow.createFlow` directly under it but it is screwing with my mind and a flow that we are retrying __should__ be a memory resident flow under all scenarios. I wonder if there is a middle ground where code is still reused without this potential confusion popping up.

williamvigorr3 replied:
Yeah this is a bit confusing. `NonResidentFlow` was originally envisioned as the only way to build a flow but actually we don't use `NonResident` flow everywhere. I think I should try and move all of this flow building code into it's own class.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -144,6 +235,12 @@ class SingleThreadedStateMachineManager(
         this.checkpointSerializationContext = checkpointSerializationContext
         this.actionExecutor = makeActionExecutor(checkpointSerializationContext)
         fiberDeserializationChecker?.start(checkpointSerializationContext)
+        when (startMode) {
```
In general im not sure if this is exactly how we want to do this. We can check with @rick-r3 but that can come after the feature generally works.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -153,6 +250,18 @@ class SingleThreadedStateMachineManager(
+        val pausedFlows = restoreFlowPrimitivesFromPausedCheckpoint()
+        mutex.locked {
+            nonResidentFlows.putAll(pausedFlows)
+            for (flow in pausedFlows) {
```
Can be done as `for((key, value) in pausedFlows)` I think that tidies it up a bit

williamvigorr3 replied:
This should be fixed.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -183,6 +292,7 @@ class SingleThreadedStateMachineManager(
     override fun stop(allowedUnsuspendedFiberCount: Int) {
         require(allowedUnsuspendedFiberCount >= 0){"allowedUnsuspendedFiberCount must be greater than or equal to zero"}
         mutex.locked {
+            if (stopped) return
```
Why do we need to add `stopped`?

williamvigorr3 replied:
So this lets me shutdown the Statemachine in the mock network tests pause a flow and then restart the mock node. Otherwise the mock network try's to shutdown the Statemachine again a throws a horrible exception. We can probably get rid of this now and just use the safe startup mode (assuming I can make the state machine start up in safe mode through the mock network somehow).

williamvigorr3 replied:
This is removed now as it is unneeded now due to this PR only supporting pausing flows when the node starts up.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -264,6 +377,56 @@ class SingleThreadedStateMachineManager(
         }
     }
 
+    override fun markFlowsAsPaused(id: StateMachineRunId): Boolean {
```
`markFlowsAsPaused` -> `markFlowAsPaused` remove the s

williamvigorr3 replied:
This is fixed

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -264,6 +377,56 @@ class SingleThreadedStateMachineManager(
     }
 
+    override fun markFlowsAsPaused(id: StateMachineRunId): Boolean {
+        mutex.locked {
```
We might not need to worry about exposing some of the functions at the moment.

williamvigorr3 replied:
Yes. Depends what we want really. Simplest thing is probably a dafe startup mode which pauses all flows on startup and an API to resume them.

rick-r3 replied:
This does have the power to pause a flow that is in the hospital, which has some use, but I think it might be worth narrowing the scope and removing this method.  We can look to bring it back as a distinct feature or PR.

rick-r3 replied:
I think we would also want to inform the hospital that we had gone from "in for observation" -> "paused".

If we are going to take action on "things in the hospital", we should have the hospital have a single thread that it processes actions on so we don't need to deal with race conditions in the hospital.

e.g. you would submit a task to the hospital to pause a flow (or retry a flow, and probably kill a flow too).

This makes it easier to deal with things in the future where many RPC clients etc are competing to do things with the flows. 

I think I'm still in favour of delaying implementation of this until we have landed the rest of this functionality?

lankydan replied:
Yeah we shouldn't expose this as its going to be a decent piece of work and needs some thorough testing.

lankydan replied:
I wonder if this could do with using the `forceRemoveFlow` function that I just removed from the kill flow work! 😄 (or a similar idea of function).

williamvigorr3 replied:
Okay I will remove this from the state machine API. For this PR we will just have a Safe Startup mode (which pauses all the flows before anything runs) and an API to resume flows.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -264,6 +377,56 @@ class SingleThreadedStateMachineManager(
+        }
+    }
+
+    override fun unPauseFlow(id: StateMachineRunId): Boolean {
```
Personally, I think unpausing a flow should be part of retrying a flow. In other words, if you retry a paused flow, it should unpause it and start the flow.

lankydan replied:
This again can be resolved when we have decided how to trigger all this stuff.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -483,9 +650,12 @@ class SingleThreadedStateMachineManager(
-                mutex.locked { flows[flowId] }?.run {
-                    fiber.scheduleEvent(Event.DeliverSessionMessage(sessionMessage, deduplicationHandler, sender))
-                } ?: logger.info("Cannot find fiber corresponding to flow ID $flowId")
+                val event = Event.DeliverSessionMessage(sessionMessage, deduplicationHandler, sender)
```
I think this is a good place for a comment about why we add external events or place it in `NonResidentFlow.externalEvents`. Maybe the second is best after thinking about it :)

lankydan replied:
Also, try a `when` block and see if that looks any better. I think this looks ok but maybe `when` looks better.

williamvigorr3 replied:
I think what we have looks better than:
```
when (flowId) {
          in flows -> { flows[flowId]!!.fiber.scheduleEvent(event) }
          in nonResidentFlows -> {nonResidentFlows[flowId]!!.addExternalEvent(event)}
          else -> logger.info("Cannot find fiber corresponding to flow ID $flowId")
}
```

lankydan replied:
👍

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -338,17 +501,23 @@ class SingleThreadedStateMachineManager(
-                )
+                val checkpoint = tryDeserializeCheckpoint(serializedCheckpoint, id) ?: return@mapNotNull null
+                val flowPrimative = NonResidentFlow(id, checkpoint, true, false, null)
+                flowPrimative.createFlow()
```
Same comment as before about it possibly being confusing here due to `NonResidentFlow` then calling `createFlow` straight after.

williamvigorr3 committed:
Minor reformating in test

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -264,6 +377,56 @@ class SingleThreadedStateMachineManager(
+
+    override fun unPauseFlow(id: StateMachineRunId): Boolean {
+        mutex.locked {
+            val flowPrimitive = nonResidentFlows[id] ?: return false
```
Think you can call `remove` which will returns `true` if it was in the map. Although that might change the semantics of this function.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,95 @@ class SingleThreadedStateMachineManager(
+            externalEvents.add(message)
+        }
+
+        fun createFlow() : Flow? {
```
Maybe some of these methods can go onto an abstract class or maybe put into `Flow`? That might remove some of my confusion due to references to `NonResidentFlow` is some of the other comments.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,95 @@ class SingleThreadedStateMachineManager(
+            }
+        }
+
+        private fun createTransientValues(id: StateMachineRunId, resultFuture: CordaFuture<Any?>): FlowStateMachineImpl.TransientValues {
```
This method is duplicated.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -272,6 +272,13 @@ class DBCheckpointStorage(
         currentDBSession().update(updateDBCheckpoint(id, checkpoint, serializedFlowState))
     }
 
+    override fun markCheckpointAsPaused(id: StateMachineRunId) : Boolean {
```
I know we will hit problems with this as we are doing performance tuning to remove calls get `get`. Therefore I think it would be best to write a direct update statement here since you have everything you need to do so.

williamvigorr3 replied:
This should never be called by any performance critical code (I suppose bring up a node in safe mode might be slow if this is really really terrible). Can write a direct update SQL statement if you want? You and @kyriathar have more of a understanding of this than me.

lankydan replied:
Yeah we should write a query. We might want to a different version of the function that takes a list/set of ids so we can bulk update. This might require a bit of rewriting in `SingleThreadedStateMachineManager` but will allow us to pause everything with 1 statement instead of many.

williamvigorr3 committed:
Correctly set anyCheckpointPersisted when retrying on error.

This caused one of the tests to lock up.

williamvigorr3 committed:
Set in memory FlowStatus back to Runnable when a checkpoint is run

I accidentally changed this behavour.

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/FlowPausingTest.kt:
```diff
@@ -0,0 +1,113 @@
+        driver(DriverParameters(startNodesInProcess = true, inMemoryDB = false)) {
+            val alice = startNode(NodeParameters(providedName = ALICE_NAME, rpcUsers = listOf(rpcUser))).getOrThrow()
+            val bob = startNode(NodeParameters(providedName = BOB_NAME, rpcUsers = listOf(rpcUser)))
+            val startedBob = bob.getOrThrow()
```
Really minor one, val `bob` is never used apart from here, so would it be better to just use one val.

williamvigorr3 replied:
Yes.

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/FlowPausingTest.kt:
```diff
@@ -0,0 +1,113 @@
+        }
+    }
+
+    fun CordaRPCOps.waitForFlowToStart(maxTrys: Int): StateMachineRunId? {
```
Would it be better to change the logic of this, to avoid `Thread.sleep`(s). Like for example since we are using in process nodes, then synchronize threads with some `Semaphore` .

williamvigorr3 replied:
So i initially tried this approach but I couldn't get it to work. Probably worth revisiting now quickly (I am not sure if I was running the test in process at that point).

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/FlowPausingTest.kt:
```diff
@@ -0,0 +1,113 @@
+            // We wait here for the initiated flow to start running on bob
+            val initiatedFlowId = startedBob.rpc.waitForFlowToStart(150)
+            assertNotNull(initiatedFlowId)
+            /* We shut down bob, we want this to happen before bob has finished receiving all of the heartbeats.
```
Very nicely identified, could we do something to avoid a potentially flaky test? like make bob's thread wait on some `Semaphore` under condition.

williamvigorr3 committed:
Fixed tests broken by rebasing

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/FlowPausingTest.kt:
```diff
@@ -0,0 +1,113 @@
+
+            //Sleep for long enough so BOB has time to receive all the messages.
+            //All messages in this period should be queued up and replayed when the flow is unpaused.
+            Thread.sleep(TOTAL_MESSAGES * SLEEP_BETWEEN_MESSAGES_MS)
```
I think this could also be synced; Alice's flow, once it sends all its messages it could release a `Semaphore `and then, the same `Semaphore` could be acquired here instead.

williamvigorr3 replied:
So yes ideally all of this test would use synchronisation between the test thread and the fibre.

kyriathar commented on node/src/integration-test/kotlin/net/corda/node/services/statemachine/FlowPausingTest.kt:
```diff
@@ -0,0 +1,113 @@
+class FlowPausingTest {
+
+    companion object {
+        val TOTAL_MESSAGES = 100
```
These static vals perhaps, if say we end up using them only inside the flow pair, they could be moved to an object containing the two flows.

williamvigorr3 replied:
Yeah that is probably better, maybe one day we will have lots of tests here.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowPausingTests.kt:
```diff
@@ -0,0 +1,188 @@
+        assertEquals(true, restartedAlice.smm.unPauseFlow(flow.id))
+        assertEquals(false, restartedAlice.smm.unPauseFlow(flow.id))
+        assertEquals(1, restartedAlice.smm.snapshot().size)
+        Thread.sleep(2000) //Forgive Me
```
Same as above.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowPausingTests.kt:
```diff
@@ -0,0 +1,188 @@
+        assertEquals(0, restartedAlice.smm.snapshot().size)
+        assertEquals(true, restartedAlice.smm.unPauseFlow(flow.id))
+        assertEquals(1, restartedAlice.smm.snapshot().size)
+        Thread.sleep(2000) //Forgive Me
```
Could we possibly get the single flow returned from restartedAlice.smm.snapshot() and wait on its future instread ?

williamvigorr3 replied:
The future no longer finishes if the node is restarted and the flow paused in the middle. unPausing a flow could in theory return a new future but I don't think we want that in terms of the API.

kyriathar replied:
Yes but haven't we unpaused the flow right above ?

williamvigorr3 replied:
Yes but the original future doesn't work anymore (even in this case where everything is done in the same process).  I am not sure if it is possible to make the future still work in a real node. It doesn't seem worth wiring it up just for the in process case.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowPausingTests.kt:
```diff
@@ -0,0 +1,188 @@
+        }
+        val restartedAlice = restartNode(aliceNode, StateMachineManager.StartMode.Safe)
+        assertEquals(0, restartedAlice.smm.snapshot().size)
+        Thread.sleep(10000) //Forgive Me
```
I m not sure but, this sleep might not be needed at all. Since we running the node as an 'in process node', then the testing thread will pause all flows.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowPausingTests.kt:
```diff
@@ -0,0 +1,188 @@
+
+        val restartedAlice = restartNode(aliceNode, StateMachineManager.StartMode.PauseHospitalised)
+        assertEquals(flows.size, restartedAlice.smm.snapshot().size)
+        Thread.sleep(10000) //Forgive Me
```
Same as above.

kyriathar commented on node/src/test/kotlin/net/corda/node/services/statemachine/FlowPausingTests.kt:
```diff
@@ -0,0 +1,188 @@
+    }
+
+    fun StateMachineManager.waitForFlowToBeHospitalised(id: StateMachineRunId) : Boolean {
+        for (i in 0..100) {
```
If 'wait for a flow to be hospitalized' means we want to wait until the flow got past the flow hospital in memory structure, then there are handy hooks there where we could attached some hook containing a `Semaphore`. If we want to wait until it stored its `HOSPITALIZED` status in DB then we don't have such place to hook at.

williamvigorr3 replied:
So yeah often want to wait until it hits the DB otherwise, the overnight observation transition might change the status from `PAUSED` to `HOSPITALIZED` if this happens then the statemachine will try and start the flow again when it restarts.

williamvigorr3 committed:
Remove safe startup mode pause hospitalised

williamvigorr3 committed:
Rename API and fix a broken test.

Rename "getRunnableCheckpoints" to "getCheckpointsToRun" as
we don't want to get the paused checkpoints separately. Also fix
a test which relied on the previous behaviour.

williamvigorr3 committed:
Improve tests use future instead of sleeps and add comments

williamvigorr3 committed:
Minor improvements to SingleThreadedStateMachineManager

Improve code quality no functional changes.

williamvigorr3 committed:
Add comment on list of events to be processed by paused flows resume.

williamvigorr3 committed:
Minor non functional refactor of SingleThreadedStateMachineManager

williamvigorr3 committed:
Add commandline option to start up node in safe mode.

This prevents the statemachine from running checkpointed flows
when in this mode.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,97 @@ class SingleThreadedStateMachineManager(
+
+        val externalEvents = mutableListOf<Event.DeliverSessionMessage>()
+
+        val checkpoint = oldCheckpoint.copy(status = Checkpoint.FlowStatus.RUNNABLE)
```
We don't want to load the checkpoint here. The point of the pause work is to start the node without loading the possibly broken data held in `Checkpont.flowState`.

lankydan replied:
I think what we need to do is:

- Load the `CheckpointState` directly (or maybe loading `Checkpoint.Serialized` without deserializing everything is enough).
- When we move from paused -> running we need to load the rest of the `Checkpoint` (just load it all again) from the database.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,97 @@ class SingleThreadedStateMachineManager(
+            return Flow(fiber, resultFuture)
+        }
+
+        private fun getFibreFromCheckpoint(): FlowStateMachineImpl<*>? {
```
Let's spell this as `Fiber` to keep things consistent :D

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -80,6 +80,97 @@ class SingleThreadedStateMachineManager(
+        }
+
+        fun createFlow() : Flow? {
+            val fiber = getFibreFromCheckpoint() ?: return null
```
Not sure if we just havent got around to this one yet or not.

Still not that sure about the naming of this function. Maybe something like `toRunnableFlow` or `toResidentFlow` would be better.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -310,7 +317,7 @@ class DBCheckpointStorage(
     }
 
-    override fun getRunnableCheckpoints(): Stream<Pair<StateMachineRunId, Checkpoint.Serialized>> {
+    override fun getCheckpointsToRun(): Stream<Pair<StateMachineRunId, Checkpoint.Serialized>> {
```
Why did we rename this, I thought `getRunnableCheckpoints` was a good name.

williamvigorr3 replied:
So I thought this was a bit confusing as `paused` checkpoints are technically 'Runnable'. I changed the behaviour here.

lankydan replied:
Ah I see. Let's have a think about it when we have addressed other stuff.

lankydan replied:
thats annoying since `getRunnable...` sounds better but you are right that it could be confusing.

williamvigorr3 committed:
Begin pulling out flows creation from SingleThreadedStateMachineManager

This should hopefully make things a lot cleaner.

williamvigorr3 committed:
Remove accidentally commited code

rick-r3 commented on node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:
```diff
@@ -48,6 +49,12 @@ open class SharedNodeCmdLineOptions {
 
+    @Option(
+            names = ["--safe-mode"],
+            description = ["Do not run any flows on startup. Flows can be resumed via RCP."]
```
RCP -> RPC

rick-r3 replied:
Maybe explicitly say "sets all flows to paused" as well.   And "Flows can be unpaused via RPC", since that is the API.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -264,6 +306,46 @@ class SingleThreadedStateMachineManager(
+    override fun markAllFlowsAsPaused(): Map<StateMachineRunId, Boolean> {
+        return checkpointStorage.getCheckpointsToRun().use {
+            it.map { (id, _) ->
+                val paused = markFlowAsPaused(id)
```
I think this is the method that will have a big performance issue.  We should put a "mark all as paused" method on `DBCheckpointStorage` (and not worry too much about the performance of the individual flow pause).

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt:
```diff
@@ -80,6 +85,29 @@ interface StateMachineManager {
+     *
+     * @return whether the flow was successfully marked as paused.
+     */
+    fun markFlowAsPaused(id: StateMachineRunId): Boolean
```
As mentioned on the implementation, perhaps leave this out for now.

williamvigorr3 committed:
Remove ability to PauseFlow from statemachine public API

This only worked on flows which were not running or Hospitalised

williamvigorr3 committed:
Add manual statement to extract paused checkpoints from DB

This means we no longer pull out the serialized flow state, which
should use less RAM. Added a test of this functionality.

williamvigorr3 committed:
Add SQL statement to pause all checkpoints in database.

This is currently unused. I need to update the statemachine
manager to use this + remove the old API. Also added a unit
test of the functionality.

williamvigorr3 committed:
Remove code to set a individual to paused.

williamvigorr3 committed:
When unpausing extract the serialised flow state.

This should make pausing work again.

williamvigorr3 committed:
Improve code for creating flow from checkpoint.

Tidied stuff up a lot. Made NonResidentFlow a lot simpler.

williamvigorr3 committed:
Clean up rextracting flow from DB when resuming a paused flow.

williamvigorr3 committed:
Make fetch paused checkpoints only fetch the paused checkpoints.

I don't know how I missed that.

williamvigorr3 committed:
Update test to check that FlowState becomes Completed/Paused

When FlowStatus is COMPLETED/PAUSED respectively.

williamvigorr3 committed:
Move code from startFlowInternal to FlowCreator

Try to resuse as much code as possible between creating flows from
checkpoint and a new flow.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -327,6 +336,56 @@ class DBCheckpointStorage(
+
+    override fun getPausedCheckpoints(): Stream<Pair<StateMachineRunId, Checkpoint.Serialized>> {
+        val session = currentDBSession()
+        val jpqlQuery = "select new ${DBPausedFields::class.java.name}(f.id, c.checkpoint, f.status, f.progressStep, f.ioRequestType, " +
```
Didn't know the queries can be put together like this.

Use a string block `"""` (3 speach marks).

Rename `c` to `checkpoint` and `f` to `blob`?

williamvigorr3 replied:
Yeah this seemed like the best way to just extract some but not all fields from the database.

lankydan commented on node/src/main/kotlin/net/corda/node/services/persistence/DBCheckpointStorage.kt:
```diff
@@ -327,6 +336,56 @@ class DBCheckpointStorage(
         return currentDBSession().find(DBFlowCheckpoint::class.java, id.uuid.toString())
     }
 
+    private class DBPausedFields(
```
Can we move this class further down in the code somewhere? Not sure of a good place to move it to though.

lankydan commented on node/src/main/kotlin/net/corda/node/services/api/CheckpointStorage.kt:
```diff
@@ -46,5 +52,17 @@ interface CheckpointStorage {
+     * Stream paused checkpoints from the store. If this is backed by a database the stream will be valid
+     * until the underlying database connection is closed, so any processing should happen before it is closed.
+     */
+    fun getPausedCheckpoints(): Stream<Pair<StateMachineRunId, Checkpoint.Serialized>>
```
Might be worth documenting that we don't load the flow state in this function.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -113,6 +121,7 @@ class SingleThreadedStateMachineManager(
 
     private var checkpointSerializationContext: CheckpointSerializationContext? = null
     private var actionExecutor: ActionExecutor? = null
+    private var flowCreator: FlowCreator? = null
```
Are we able to change this to a `lateinit var`?

If that works, I wonder if we can do the same for `actionExecutor`.

williamvigorr3 replied:
Yes and we can even move actionExecutor into start as it is only used inside the `FlowCreator`.

lankydan replied:
Yes, good observation.

lankydan replied:
Maybe we take it even further and consider moving some of the dependencies into the `FlowCreator`. Would that be worth it?

williamvigorr3 replied:
I just checked and nothing else can be easily moved in. More modularisation would be nice.

williamvigorr3 committed:
Clarify command line option description.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -0,0 +1,182 @@
+                val logic = tryCheckpointDeserialize(checkpoint.flowState.frozenFlowLogic, runId) ?: return null
+                FlowStateMachineImpl(runId, logic, scheduler)
+            }
+            is FlowState.Started -> {
```
Remove the `{}` for single line when blocks

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -0,0 +1,182 @@
+        return createFlow(checkpoint, flowStateMachineImpl, existingCheckpoint != null, resultFuture, deduplicationHandler, senderUUID)
+    }
+
+    private fun getFiberFromCheckpoint(runId: StateMachineRunId, checkpoint: Checkpoint): FlowStateMachineImpl<*>? {
```
This might look nicer as an extension function. `Checkpoint.getFiber`.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -0,0 +1,182 @@
+    }
+}
+
+class FlowCreator(val checkpointSerializationContext: CheckpointSerializationContext,
```
Generally can we correct the indentation in this file. The kotlin code style should have waaaaaay less identation than this.

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/FlowCreator.kt:
```diff
@@ -0,0 +1,182 @@
+                   resultFuture: OpenFuture<Any?>,
+                   deduplicationHandler: DeduplicationHandler? = null,
+                   senderUUID: String? = null): Flow<A> {
+        val state = StateMachineState(
```
Can we break out the creation of the `StateMachineState` into a function like `createTransientValues` and call it in `createFlowFromCheckpoint` and `createFlowFromLogic`?

It will increase duplication slightly but I think it gives better clarify (basically why does `createFlow` create the `transientState` but not the `transientValues`).

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/SingleThreadedStateMachineManager.kt:
```diff
@@ -333,23 +384,28 @@ class SingleThreadedStateMachineManager(
     }
 
-    private fun resumeRestoredFlows(flows: List<Flow>) {
+    private fun restoreFlowPrimitivesFromPausedCheckpoint(): Map<StateMachineRunId, NonResidentFlow> {
```
Maybe `restoreNonResidentFlowsFromPausedCheckpoints`? (either way, add a `s` to the end)

lankydan commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineState.kt:
```diff
@@ -163,9 +163,14 @@ data class Checkpoint(
          */
         fun deserialize(checkpointSerializationContext: CheckpointSerializationContext): Checkpoint {
+            val flowState = when(status) {
+                FlowStatus.PAUSED -> {FlowState.Paused}
```
the `{}` can be removed

lankydan reviewed (commented):
I think `FlowCreator` looks pretty good now. Left some tidy up comments but nothing stands out to me as wrong.

williamvigorr3 committed:
Make command line option work

williamvigorr3 committed:
Refactored flow creator to be cleaner.

Also a large amount of tidy up.

williamvigorr3 committed:
Add newline again (accidentally kept original in rebase dam Vimdiff)

williamvigorr3 committed:
Organise Imports

rick-r3 commented on core/src/main/kotlin/net/corda/core/internal/messaging/InternalCordaRPCOps.kt:
```diff
@@ -13,4 +14,11 @@ interface InternalCordaRPCOps : CordaRPCOps {
     val attachmentTrustInfos: List<AttachmentTrustInfo>
+
+    /**
+     * Start a paused flow.
```
Resume rather than start.

williamvigorr3 replied:
I have fixed this.

rick-r3 commented on node/src/main/kotlin/net/corda/node/NodeCmdLineOptions.kt:
```diff
@@ -48,6 +49,12 @@ open class SharedNodeCmdLineOptions {
     var devMode: Boolean? = null
 
+    @Option(
+            names = ["--safe-mode"],
```
Was the name of the command line option defined e.g. in the FEAT ?

I'm just wondering if "safe mode" is a bit generic, and whether that might want to be saved for something else, and a more specific "pause-all-flows" is better.  Not thought about it before so just thinking aloud.

williamvigorr3 replied:
So the name was not defined in the Feat (FEAT-51), only the CORDA ticket mentioned `safe` startup mode. Happy to go with `--pause-all-flows`. @lankydan are you also happy with that?

lankydan replied:
makes sense

williamvigorr3 replied:
I have fixed this.

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumperImpl.kt:
```diff
@@ -141,7 +141,7 @@ class CheckpointDumperImpl(private val checkpointStorage: CheckpointStorage, pri
             if (lock.getAndIncrement() == 0 && !file.exists()) {
                 database.transaction {
-                    checkpointStorage.getRunnableCheckpoints().use { stream ->
+                    checkpointStorage.getCheckpointsToRun().use { stream ->
```
I think maybe this should include all checkpoints.  i.e. not even just runnable checkpoints, but also paused or even incompatible ones.  I assume in the main branch ("the original") it just does everything ?  I think that is still valid.  You might be wanting to examine something that isn't "to run".

rick-r3 replied:
Or maybe incompatible ones don't make sense because it can't be deserialized.... but paused ones should be included I think.

williamvigorr3 replied:
Originally (the 4.5 release branch) we dumped all the checkpoints. On the feature branch we are currently excluding `FAILED`, `COMPLETED` and `HOSPITALIZED` checkpoints. Maybe this should just return all the checkpoints (except `COMPLETED)`. When if we fail to deserialise here then this will be logged by the checkpoint dumper.

rick-r3 replied:
I think all excluding `FAILED`, `COMPLETED` and anything incompatible (as I assume it wouldn't deserialize).

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/statemachine/StateMachineManager.kt:
```diff
@@ -30,12 +30,18 @@ import java.time.Duration
 interface StateMachineManager {
+
+    enum class StartMode {
+        ResumeAll, // Resume all flows except paused flows.
```
`ResumeAll` sounds like it includes paused flows to me.  Should it just be `ExcludingPaused` ?  I can imagine us adding a `ResumeAll` that unpauses everything first and then does indeed resume all (not now, maybe in the future).

williamvigorr3 replied:
I agree that is clearer so have changed this to `ExcludingPaused`.

williamvigorr3 committed:
Minor refactor to tidy some stuff up.

williamvigorr3 committed:
Make checkpoint dumper, dump paused and hospitalized flows.

Added new method to CheckpointStorage to do this + improve test.

williamvigorr3 committed:
Fix name shawdowing

williamvigorr3 committed:
Clean up API in CheckpointStorage

rick-r3 commented on node/src/main/kotlin/net/corda/node/services/rpc/CheckpointDumperImpl.kt:
```diff
@@ -141,7 +146,7 @@ class CheckpointDumperImpl(private val checkpointStorage: CheckpointStorage, pri
             if (lock.getAndIncrement() == 0 && !file.exists()) {
                 database.transaction {
-                    checkpointStorage.getRunnableCheckpoints().use { stream ->
+                    checkpointStorage.getCheckpoints(DUMPABLE_CHECKPOINTS).use { stream ->
```
Yeah, I think this works.

rick-r3 reviewed (approved):
I've realised there's probably a currently unspecified feature we should add.  Let's say we pause all the flows.  Next thing you know, a load of flows get invoked on your node from remote peers and your node is overloaded again.  There's some capability there in the form of flow draining mode.  @lankydan Can we create a JIRA to track ?  We should experiment with what is possible currently, and see if there's a gap.
